apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -338,54 +444,136 @@ export async function aiGrade({
-        response_format: zodResponseFormat(GPTGradeSchema, 'grades'),
-      });
-
-      try {
-        msg += `Number of tokens used: ${completion.usage ? completion.usage.total_tokens : 0}\n`;
-        const grade_response = completion.choices[0].message;
-        msg += `Raw ChatGPT response:\n${grade_response.content}`;
-        if (grade_response.parsed) {
-          await manualGrading.updateInstanceQuestionScore(
-            assessment_question.assessment_id,
-            instance_question.id,
-            submission.id,
-            null, // modified_at
-            {
-              score_perc: grade_response.parsed.grade,
-              feedback: { manual: grade_response.parsed.feedback },
-              // NEXT STEPS: rubrics
-            },
-            user_id,
+      if (rubric_items.length > 0) {
+        // Dynamically generate the rubric schema based on the number of items
+        let GPTRubricItemSchema = z.object({});
+        for (let i = 0; i < rubric_items.length; i++) {
+          GPTRubricItemSchema = GPTRubricItemSchema.merge(
+            z.object({
+              [i]: z.object({ description: z.string(), selected: z.boolean() }),
+            }),
+          );
+        }
+        const GPTRubricGradeSchema = z.object({
+          rubric_items: GPTRubricItemSchema,
+          feedback: z.string(),
+        });
+        const completion = await openai.beta.chat.completions.parse({
+          messages,
+          model: OPEN_AI_MODEL,
+          user: `course_${course.id}`,
+          response_format: zodResponseFormat(GPTRubricGradeSchema, 'grades'),
+        });
+        try {
+          job.info(
+            `Number of tokens used: ${completion.usage ? completion.usage.total_tokens : 0}`,
           );
-          msg += `\nAI grades: ${grade_response.parsed.grade}`;
-        } else if (grade_response.refusal) {
+          const grade_response = completion.choices[0].message;
+          job.info(`Raw ChatGPT response:\n${grade_response.content}`);
+          if (grade_response.parsed) {
+            // Only care about the rubric numbers
+            const gptRubricItems: number[] = [];
+            for (let i = 0; i < rubric_items.length; i++) {
+              const item = grade_response.parsed.rubric_items[i];
+              if (item.selected) {
+                gptRubricItems.push(i);
+              }
+            }
+            const manual_rubric_data = {
+              rubric_id: rubric_items[0].rubric_id,
+              applied_rubric_items: parseRubricItems({
+                rubric_items,
+                gptRubricItems,
+              }),
+            };
+            new_rubric_items = await queryRows(
+              sql.select_rubric_for_grading,
+              {
+                assessment_question_id: assessment_question.id,
+              },
+              RubricItemSchema,
+            );
+            // Check if rubric items has been modified
+            validateRubric(rubric_items, new_rubric_items);
```

Comment:
For a function whose sole job is usually to throw an error, I typically name it `assert...`, e.g. `assertRubricNotModified(...)`. IMO this is a little clearer more self-documenting.